doc/proposals/fast_ipam_by_node.md

```diff
@@ -0,0 +1,268 @@
+      RangeStart          net.IP               `json:"range_start,omitempty"`
+      RangeEnd            net.IP               `json:"range_end,omitempty"`
+      GatewayStr          string               `json:"gateway"`
+      EtcdHost            string               `json:"etcd_host,omitempty"`
+      EtcdUsername        string               `json:"etcd_username,omitempty"`
+      EtcdPassword        string               `json:"etcd_password,omitempty"`
+      EtcdKeyFile         string               `json:"etcd_key_file,omitempty"`
+      EtcdCertFile        string               `json:"etcd_cert_file,omitempty"`
+      EtcdCACertFile      string               `json:"etcd_ca_cert_file,omitempty"`
+      LeaderLeaseDuration int                  `json:"leader_lease_duration,omitempty"`
+      LeaderRenewDeadline int                  `json:"leader_renew_deadline,omitempty"`
+      LeaderRetryPeriod   int                  `json:"leader_retry_period,omitempty"`
+      LogFile             string               `json:"log_file"`
+      LogLevel            string               `json:"log_level"`
+      OverlappingRanges   bool                 `json:"enable_overlapping_ranges,omitempty"`
+      SleepForRace        int                  `json:"sleep_for_race,omitempty"`
+      Gateway             net.IP
+      Kubernetes          KubernetesConfig     `json:"kubernetes,omitempty"`
+      ConfigurationPath   string               `json:"configuration_path"`
+      PodName             string
+      PodNamespace        string
+ }
+```
+
+```diff
+type PoolIdentifier struct {
+	IpRange     string
+	NetworkName string
++	NodeName    string # this will signal that fast node slicing is enabled
+}
+
+func IPPoolName(poolIdentifier PoolIdentifier) string {
+-	if poolIdentifier.NetworkName == UnnamedNetwork {
+-		return normalizeRange(poolIdentifier.IpRange)
++	if poolIdentifier.NodeName != "" {
++		// fast node range naming convention
++		if poolIdentifier.NetworkName == UnnamedNetwork {
++			return fmt.Sprintf("%v-%v", normalizeRange(poolIdentifier.IpRange), poolIdentifier.NodeName)
++		} else {
++			return fmt.Sprintf("%v-%v", poolIdentifier.NetworkName, poolIdentifier.NodeName)
++		}
+	} else {
+-		return fmt.Sprintf("%s-%s", poolIdentifier.NetworkName, normalizeRange(poolIdentifier.IpRange))
++		// default naming convention
++		if poolIdentifier.NetworkName == UnnamedNetwork {
++			return normalizeRange(poolIdentifier.IpRange)
++		} else {
++			return fmt.Sprintf("%s-%s", poolIdentifier.NetworkName, normalizeRange(poolIdentifier.IpRange))
++		}
+	}
+}
+
+```
+
+Additional changes will be required within whereabouts to use the NodeSlice to find the `Range` that the node its running
+on is assigned to. From here it can use the range on the IPPool with current code. Another change to set the IPPoolName and 
+Lease name as described above will be required. Finally, a new controller will be introduced to assign nodes to NodeSlices.
```

Comment:
I'm definitely a-ok with the new controller. I think it's the right way forward.